### PR TITLE
Pub/Sub publish callbacks with original Message awareness

### LINKED ISSUE #467: PubSub Message Handler Async At-least-once Delivery
**Problem**

There is no way to ship the data into GCP PubSub via using com.google.cloud.spring.pubsub.integration.outbound.PubSubMessageHandler & with meeting as follows requirements:
* do it asynchronously (with batching is plus)
* be able programatically achieve finally at-least-once delivery guarantee

To be more descriptive here:
* configure PubSubMessageHandler in async mode
* in this case there is only "publish callback" to be provided for operation status notification
* "publish callback" provides only GCP PubSub message ID as input parameter
* so in "publish callback" original input message is missed + headers
* as result there is no way to correlate original input message (submitted into PubSubMessageHandler) with GCP PubSub response
* as result no way to mark original input (source) message as delivered or redeliver it if something happened

        PubSubMessageHandler adapter = new PubSubMessageHandler(pubsubTemplate, topic);

        adapter.setSync(true);

        adapter.setPublishCallback(new ListenableFutureCallback<String>() {
            @Override
            public void onFailure(Throwable ex) {
                log.debug("There was an error sending the message", ex);
            }

            @Override
            public void onSuccess(String result) {
                log.debug("Message was sent successfully - result: {}", result);
            }
        });

And would like to emphasise that trying to achieve at-least-once delivery guarantee. So "sigkill" should be handled by my app without data loss.

Consequently, at the moment custom PubSubMessageHandler implementation is required to achieve what was described above. Or please correct me if I've missed something. 
 
**Versions**
* spring-cloud.version --> 2020.0.2
* spring-cloud-gcp.version --> 2.0.2

**Describe the solution you'd like**
Do similar as in org.springframework.integration.http.outbound.HttpRequestExecutingMessageHandler:
* add possibility to specify "OutputChannel"
* in a message delegated into "OutputChannel" preserve message headers of original input message

Having this will be a possibility to correlate PubSub response with input source message. And mark/rollback it if needed.

### PR DESCRIPTION
**Pull request by elefeint**
Two callbacks provided: `setSuccessCallback()` and `setFailureCallback()`. 
Second parameter to both is the original `Message` for consistency, and to enable forwarding to downstream channels.

Fixes #467.

### CODE REVIEW

File: spring-cloud-gcp-pubsub/src/main/java/com/google/cloud/spring/pubsub/integration/outbound/PubSubMessageHandler.java [hunk 1]
```diff
@@ -130,11 +134,31 @@ public void setPublishTimeout(long timeoutMillis) {
 	/**
 	 * Set the callback to be activated when the publish call resolves.
 	 * @param publishCallback callback for the publish future
+	 * @deprecated Use {@code setSuccessCallback()} and {@code setFailureCallback()} instead.
```
Comment (meltsufin) [inline spring-cloud-gcp-pubsub/src/main/java/com/google/cloud/spring/pubsub/integration/outbound/PubSubMessageHandler.java:137]: You can use `{@link` here.

File: spring-cloud-gcp-pubsub/src/main/java/com/google/cloud/spring/pubsub/integration/outbound/PubSubMessageHandler.java [hunk 2]
```diff
@@ -130,11 +134,31 @@ public void setPublishTimeout(long timeoutMillis) {
...
 	 */
+	@Deprecated
 	public void setPublishCallback(ListenableFutureCallback<String> publishCallback) {
 		this.publishCallback = publishCallback;
 	}
 
+	/**
+	 * Set callback (can be a lambda) for processing the published message ID and the original {@code Message} after the
+	 * message was successfully published.
+	 * @param cb callback accepting a {@code String} message ID and the original {@code Message}.
```
Comment (meltsufin) [inline spring-cloud-gcp-pubsub/src/main/java/com/google/cloud/spring/pubsub/integration/outbound/PubSubMessageHandler.java:147]: `cb` seems to be different from the long-form style. `successCallback`?

File: spring-cloud-gcp-pubsub/src/main/java/com/google/cloud/spring/pubsub/integration/outbound/PubSubMessageHandler.java [hunk 3]
```diff
@@ -130,11 +134,31 @@ public void setPublishTimeout(long timeoutMillis) {
...
+	public void setSuccessCallback(SuccessCallback cb) {
+		this.successCallback = cb;
+	}
+
+	/**
+	 * Set callback (can be a lambda) for processing the root cause exception and the original {@code Message} in case
+	 * of failure.
+	 * @param cb callback accepting a {@code Throwable} and a {@code Message}.
+	 */
+	public void setFailureCallback(FailureCallback cb) {
```
Comment (meltsufin) [inline spring-cloud-gcp-pubsub/src/main/java/com/google/cloud/spring/pubsub/integration/outbound/PubSubMessageHandler.java:158]: `failureCallback`?

File: docs/src/main/asciidoc/spring-integration-pubsub.adoc [hunk 4]
```diff
@@ -262,8 +262,10 @@ The provided `PubSubTemplate` contains all the necessary configuration to publis
 A publish timeout can be configured for synchronous publishing.
 If none is provided, the adapter waits indefinitely for a response.
 
-It is possible to set user-defined callbacks for the `publish()` call in `PubSubMessageHandler` through the `setPublishFutureCallback()` method.
-These are useful to process the message ID, in case of success, or the error if any was thrown.
+It is possible to set user-defined callbacks for the `publish()` call in `PubSubMessageHandler` through the `setSuccessCallback()` and `setFailureCallback()` methods (either one or both may be set).
+These give access to the Pub/Sub publish message ID in case of success, or the root cause exception in case of error.
```
Comment (meltsufin) [inline docs/src/main/asciidoc/spring-integration-pubsub.adoc:266]: "message ID" always seems confusing to me. What we're really talking about is the "message ack ID", right?

**Review by meltsufin — APPROVED** [review]
Looks good, just some minor naming comments.

File: docs/src/main/asciidoc/spring-integration-pubsub.adoc (hunk 4 above)
Comment (elefeint) [reply docs/src/main/asciidoc/spring-integration-pubsub.adoc:266]: The confusion goes down to proto level. Publish operation returns [server-assigned message ID](https://github.com/googleapis/googleapis/blob/6afe7923cbf18ac9de4163fe1af316c74540b46b/google/pubsub/v1/pubsub.proto#L277). But when we acknowledge, it's called ["ack ID"](https://github.com/googleapis/googleapis/blob/6afe7923cbf18ac9de4163fe1af316c74540b46b/google/pubsub/v1/pubsub.proto#L1056).

Comment (meltsufin) [reply docs/src/main/asciidoc/spring-integration-pubsub.adoc:266]: When you pull a message, you get an ack ID with it, which you can use to acknowledge it.
I don't think you get anything back when you acknowledge a message.

Comment (elefeint) [reply docs/src/main/asciidoc/spring-integration-pubsub.adoc:266]: Right, but we are not pulling a message. We are _publishing_ a message, so we are getting that message's ID back.

**Pull request merged**